File: app/src/host/statusLineWatcher.ts
import * as fs from 'fs';
import * as os from 'os';
import * as path from 'path';
import { createHash } from 'crypto';
import type { StatusLineSnapshot } from './types';
import { applyResetWindow } from '../lib/limit-window';
import { planLimitsBroadcast } from '../lib/limit-broadcast';

/** Shared file holding the account's rate limits, written by whichever tab last saw them. */
const LIMITS_FILE = 'limits.json';
/** Upper bound on how long a tab shows a limit another tab has already superseded. */
const LIMITS_POLL_MS = 30_000;

/**
 * Directory the statusLine script writes its per-tab snapshots into.
 * `os.tmpdir()` is per-user on macOS (`/var/folders/…`); the script additionally
 * writes with `umask 077`.
 */
export function getStatusLineDir(): string {
  return path.join(os.tmpdir(), 'figma-claude-panel', 'status');
}

export type StatusLineCallback = (terminalId: string, snapshot: StatusLineSnapshot | null) => void;

/** `~` for the home directory, matching what the producers write. */
function collapseHome(dir: string): string {
  const home = os.homedir();
  if (home.length === 0) return dir;
  if (dir === home) return '~';
  return dir.startsWith(home + path.sep) ? `~${dir.slice(home.length)}` : dir;
}

/**
 * File name for the remembered snapshot. Hashed rather than escaped so a path with separators or
 * odd characters cannot escape the directory, and collapsed first so `/Users/x/p` and `~/p` are
 * the same key.
 */
function hashCwd(cwd: string): string {
  return createHash('sha1').update(collapseHome(cwd)).digest('hex').slice(0, 16);
}

/**
 * Watches the status directory and reports whatever the statusLine script wrote.
 *
 * The extension host only ever sees PTY bytes, so model, token counts and rate limits
 * cannot be read from the terminal. Claude Code hands them to the configured statusLine
 * command instead, which writes them here as `<terminal id>.json`.
 */
export class StatusLineWatcher {
  private readonly dir = getStatusLineDir();
  /**
   * Last snapshot per working directory, kept across windows. Claude Code only runs the
   * statusLine command when it renders, which is after its first output — so a fresh tab would
   * show nothing for a while. This is what fills the row in the meantime, greyed out because its
   * `updatedAt` is old.
   */
  private readonly lastDir = path.join(getStatusLineDir(), 'last');
  private watcher: fs.FSWatcher | undefined;
  /**
   * Second watcher, on `last/`. The rate limits belong to the account, so a tab that renders
   * nothing itself can still learn a fresher percentage from one that does — in this window or
   * another. `fs.watch` is not recursive, so the directory above does not report this.
   */
  private limitsWatcher: fs.FSWatcher | undefined;
  private limitsPoll: NodeJS.Timeout | undefined;
  /** Guards against re-sending the same file: `updatedAt` of what was last broadcast. */
  private lastBroadcastLimitsAt: number | undefined;
  private readonly debounceTimers = new Map<string, NodeJS.Timeout>();
  private readonly latest = new Map<string, StatusLineSnapshot>();
  private disposed = false;

  constructor(
    private readonly onSnapshot: StatusLineCallback,
    private readonly debounceMs = 150
  ) {
    this.start();
  }

  /** The last snapshot seen for a tab, so a webview reload can be filled in again. */
  get(terminalId: string): StatusLineSnapshot | undefined {
    return this.latest.get(terminalId);
  }

  /**
   * What a brand new tab in this directory should show until Claude renders for the first time:
   * the previous session's snapshot if there is one, otherwise just the directory.
   */
  getInitialSnapshot(cwd: string | undefined): StatusLineSnapshot | undefined {
    if (cwd === undefined) {
      return undefined;
    }

    const base: StatusLineSnapshot = this.readLastForCwd(cwd) ?? {
      // Nothing remembered for this directory — the path alone still beats an empty row, and it
      // keeps the terminal height from changing once the real values arrive.
      model: '',
      cwd: collapseHome(cwd),
      usedTokens: 0,
      totalTokens: 0,
      usedPercent: 0,
      updatedAt: 0
    };

    return this.withRememberedLimits(base);
  }

  /**
   * Fills in rate limits the snapshot does not carry. Two reasons it may not:
   *
   * - Claude Code leaves `rate_limits` out of its payload until a request has been made, so the
   *   first snapshots of a session have no Session and Week at all.
   * - A brand new tab starts from the remembered directory snapshot, which may predate them.
   *
   * The limits belong to the account rather than a directory, hence their own file. Only
   * undefined fields are filled — a live value always wins over a remembered one.
   */
  private withRememberedLimits(snapshot: StatusLineSnapshot): StatusLineSnapshot {
    const limits = this.readLimits();
    if (!limits) {
      return snapshot;
    }

    const merged = { ...snapshot };
    for (const [key, value] of Object.entries(limits) as [
      keyof StatusLineSnapshot,
      number | string | undefined
    ][]) {
      if (value !== undefined && merged[key] === undefined) {
        Object.assign(merged, { [key]: value });
      }
    }

    // A remembered "resets in 84 min" is a lie an hour later; the rules live in one tested place.
    return applyResetWindow(merged, Date.now());
  }

  /**
   * Supplies the context window size when the payload has none yet, so the bar has a scale from
   * the start. The **usage** is never taken from memory: a fresh session really is at zero, and
   * showing the previous session's tokens would be a lie rather than a placeholder.
   */
  private withRememberedWindowSize(snapshot: StatusLineSnapshot): StatusLineSnapshot {
    if (snapshot.totalTokens > 0 || snapshot.cwd === undefined) {
      return snapshot;
    }

    const remembered = this.readLastForCwd(snapshot.cwd);
    if (!remembered || remembered.totalTokens <= 0) {
      return snapshot;
    }

    const totalTokens = remembered.totalTokens;
    return {
      ...snapshot,
      totalTokens,
      usedPercent: Math.round((snapshot.usedTokens / totalTokens) * 1000) / 10
    };
  }

  private readLimits(): Partial<StatusLineSnapshot> | undefined {
    try {
      const raw = fs.readFileSync(path.join(this.lastDir, LIMITS_FILE), 'utf8');
      const parsed: unknown = JSON.parse(raw);
      if (typeof parsed !== 'object' || parsed === null) {
        return undefined;
      }
      const value = parsed as Record<string, unknown>;
      return {
        sessionPercent: numberOrUndefined(value.sessionPercent),
        sessionResetsAt: numberOrUndefined(value.sessionResetsAt),
        sessionResetsInMin: numberOrUndefined(value.sessionResetsInMin),
        weekPercent: numberOrUndefined(value.weekPercent),
        weekResetsAt: typeof value.weekResetsAt === 'string' ? value.weekResetsAt : undefined,
        // Only meaningful for the broadcast below: whether this file is newer than a tab's own
        // snapshot. `withRememberedLimits` never overwrites a defined field, so it stays put.
        updatedAt: numberOrUndefined(value.updatedAt)
      };
    } catch {
      return undefined;
    }
  }

  private rememberLimits(snapshot: StatusLineSnapshot): void {
    if (snapshot.sessionPercent === undefined && snapshot.weekPercent === undefined) {
      return;
    }
    this.writeAtomically(LIMITS_FILE, {
      sessionPercent: snapshot.sessionPercent,
      sessionResetsAt: snapshot.sessionResetsAt,
      sessionResetsInMin: snapshot.sessionResetsInMin,
      weekPercent: snapshot.weekPercent,
      weekResetsAt: snapshot.weekResetsAt,
      updatedAt: snapshot.updatedAt
    });
  }

  private readLastForCwd(cwd: string): StatusLineSnapshot | undefined {
    try {
      const raw = fs.readFileSync(path.join(this.lastDir, `${hashCwd(cwd)}.json`), 'utf8');
      return parseSnapshot(raw);
    } catch {
      return undefined;
    }
  }

  private rememberForCwd(snapshot: StatusLineSnapshot): void {
    if (snapshot.cwd === undefined || snapshot.cwd.length === 0 || snapshot.totalTokens <= 0) {
      return;
    }
    this.writeAtomically(`${hashCwd(snapshot.cwd)}.json`, snapshot);
  }

  private writeAtomically(fileName: string, payload: unknown): void {
    try {
      fs.mkdirSync(this.lastDir, { recursive: true, mode: 0o700 });
      const target = path.join(this.lastDir, fileName);
      const temp = `${target}.tmp`;
      fs.writeFileSync(temp, JSON.stringify(payload), { mode: 0o600 });
      fs.renameSync(temp, target);
    } catch {
      // Remembering is a convenience; failing at it must not affect the live row
    }
  }

  /** Drops a tab's snapshot and its file. */
  removeTerminal(terminalId: string): void {
    this.clearTimer(terminalId);
    this.latest.delete(terminalId);
    try {
      fs.unlinkSync(path.join(this.dir, `${terminalId}.json`));
    } catch {
      // Never written, or already gone
    }
  }

  dispose(): void {
    this.disposed = true;
    this.watcher?.close();
    this.watcher = undefined;
    this.limitsWatcher?.close();
    this.limitsWatcher = undefined;
    if (this.limitsPoll) {
      clearInterval(this.limitsPoll);
      this.limitsPoll = undefined;
    }
    for (const terminalId of [...this.latest.keys()]) {
      this.removeTerminal(terminalId);
    }
    for (const timer of this.debounceTimers.values()) {
      clearTimeout(timer);
    }
    this.debounceTimers.clear();
  }

  private start(): void {
    try {
      fs.mkdirSync(this.dir, { recursive: true, mode: 0o700 });
    } catch (error) {
      console.warn('[Claude Terminal] status line directory unavailable', error);
      return;
    }

    // Leftovers from a previous window describe tabs that no longer exist.
    this.removeStaleFiles();

    try {
      this.watcher = fs.watch(this.dir, { persistent: false }, (_event, filename) => {
        if (filename) {
          this.scheduleRead(filename);
        }
      });
    } catch (error) {
      console.warn('[Claude Terminal] cannot watch the status line directory', error);
    }

    this.startLimitsWatch();
  }

  /**
   * Keeps the rate limits current in tabs that render nothing themselves. The percentage cannot be
   * recomputed the way the countdown can — it exists only in Claude's payload — but it is an
   * account-wide number, so whichever tab last saw one shares it through `last/limits.json`.
   *
   * Both a watch and a poll: the watch reacts at once, the interval is the guarantee, because
   * `fs.watch` is platform-dependent and an atomic rename can slip past it.
   */
  private startLimitsWatch(): void {
    try {
      fs.mkdirSync(this.lastDir, { recursive: true, mode: 0o700 });
      this.limitsWatcher = fs.watch(this.lastDir, { persistent: false }, (_event, filename) => {
        if (filename === LIMITS_FILE) {
          this.scheduleLimitsBroadcast();
        }
      });
    } catch (error) {
      console.warn('[Claude Terminal] cannot watch the remembered limits', error);
    }

    this.limitsPoll = setInterval(() => {
      this.broadcastLimits();
    }, LIMITS_POLL_MS);
    // Never hold the app open for a status row
    this.limitsPoll.unref();
  }

  /** The file is written through a temp file plus rename, so one update fires several events. */
  private scheduleLimitsBroadcast(): void {
    if (this.disposed) return;

    const existing = this.debounceTimers.get(LIMITS_FILE);
    if (existing) {
      clearTimeout(existing);
    }
    this.debounceTimers.set(
      LIMITS_FILE,
      setTimeout(() => {
        this.debounceTimers.delete(LIMITS_FILE);
        this.broadcastLimits();
      }, this.debounceMs)
    );
  }

  /** Hands newer limits to every tab whose own snapshot predates them. Rules in `limit-broadcast`. */
  private broadcastLimits(): void {
    if (this.disposed) return;

    const plan = planLimitsBroadcast(
      this.readLimits(),
      this.lastBroadcastLimitsAt,
      this.latest,
      Date.now()
    );
    if (!plan) {
      return;
    }

    this.lastBroadcastLimitsAt = plan.limitsAt;
    for (const [terminalId, merged] of plan.updates) {
      this.latest.set(terminalId, merged);
      this.onSnapshot(terminalId, merged);
    }
  }

  /**
   * Snapshots from a previous window describe tabs that no longer exist. The `last` directory is
   * deliberately kept — that is the memory a fresh tab starts from.
   */
  private removeStaleFiles(): void {
    try {
      for (const entry of fs.readdirSync(this.dir, { withFileTypes: true })) {
        if (entry.isFile()) {
          fs.unlinkSync(path.join(this.dir, entry.name));
        }
      }
    } catch {
      // Empty or unreadable — nothing to clean up
    }
  }

  /**
   * The script writes through a temp file plus rename, so a single update can fire several
   * events. Debouncing per file keeps that down to one read.
   */
  private scheduleRead(filename: string): void {
    if (this.disposed || !filename.endsWith('.json')) {
      return;
    }

    const terminalId = filename.slice(0, -'.json'.length);
    this.clearTimer(terminalId);
    this.debounceTimers.set(
      terminalId,
      setTimeout(() => {
        this.debounceTimers.delete(terminalId);
        this.read(terminalId);
      }, this.debounceMs)
    );
  }

  private read(terminalId: string): void {
    if (this.disposed) return;

    let raw: string;
    try {
      raw = fs.readFileSync(path.join(this.dir, `${terminalId}.json`), 'utf8');
    } catch {
      // File removed between event and read
      this.latest.delete(terminalId);
      this.onSnapshot(terminalId, null);
      return;
    }

    const snapshot = parseSnapshot(raw);
    if (!snapshot) {
      return;
    }

    // Remember first, from the live values, then hand out a snapshot topped up with whatever
    // rate limits are known — Claude omits them until the session has made a request.
    this.rememberForCwd(snapshot);
    this.rememberLimits(snapshot);

    const complete = this.withRememberedLimits(this.withRememberedWindowSize(snapshot));
    this.latest.set(terminalId, complete);
    this.onSnapshot(terminalId, complete);
  }

  private clearTimer(terminalId: string): void {
    const timer = this.debounceTimers.get(terminalId);
    if (timer) {
      clearTimeout(timer);
      this.debounceTimers.delete(terminalId);
    }
  }
}

/**
 * Parses a snapshot defensively. The file is written by a shell script, so a truncated or
 * half-formatted write must not take the panel down — it is dropped and the next write wins.
 */
function parseSnapshot(raw: string): StatusLineSnapshot | undefined {
  let parsed: unknown;
  try {
    parsed = JSON.parse(raw);
  } catch {
    return undefined;
  }

  if (typeof parsed !== 'object' || parsed === null) {
    return undefined;
  }

  const value = parsed as Record<string, unknown>;
  const usedTokens = numberOrUndefined(value.usedTokens);
  const totalTokens = numberOrUndefined(value.totalTokens);
  const usedPercent = numberOrUndefined(value.usedPercent);

  if (usedTokens === undefined || totalTokens === undefined || usedPercent === undefined) {
    return undefined;
  }

  return {
    model: typeof value.model === 'string' ? value.model : '',
    effort: typeof value.effort === 'string' ? value.effort : undefined,
    cwd: typeof value.cwd === 'string' ? value.cwd : undefined,
    usedTokens,
    totalTokens,
    usedPercent,
    sessionPercent: numberOrUndefined(value.sessionPercent),
    sessionResetsAt: numberOrUndefined(value.sessionResetsAt),
    sessionResetsInMin: numberOrUndefined(value.sessionResetsInMin),
    weekPercent: numberOrUndefined(value.weekPercent),
    weekResetsAt: typeof value.weekResetsAt === 'string' ? value.weekResetsAt : undefined,
    compacted: numberOrUndefined(value.compacted),
    compactBudget: numberOrUndefined(value.compactBudget),
    compactAuto: numberOrUndefined(value.compactAuto),
    updatedAt: numberOrUndefined(value.updatedAt) ?? 0
  };
}

function numberOrUndefined(value: unknown): number | undefined {
  return typeof value === 'number' && Number.isFinite(value) ? value : undefined;
}

File: swift-host/Sources/FigmaClaudeCore/StatusLineWatcher.swift
import Foundation

/// Reads back what the status line producer writes.
///
/// Port of `app/src/host/statusLineWatcher.ts`. Polling rather than FSEvents on the directory:
/// the producer writes a temporary file and renames it over the old one, and a directory watch
/// reports that inconsistently — a rename is a create plus a delete plus, on some volumes,
/// nothing at all. A stat every second is cheap and never misses one.
public final class StatusLineWatcher {
    private let dir: String
    private let interval: TimeInterval
    private let onChange: (String, StatusLineSnapshot) -> Void
    private let queue = DispatchQueue(label: "de.designdone.figmaclaude.statusline")
    private var timer: DispatchSourceTimer?
    private var seen: [String: Date] = [:]
    private var snapshots: [String: StatusLineSnapshot] = [:]
    /// What survives a tab, a window and a restart: the account's limits and the last snapshot per
    /// working directory. Port of the `last/` directory in `app/src/host/statusLineWatcher.ts:57`.
    private var lastDir: String { rememberedDir(dir) }

    public init(dir: String = statusLineDir(), interval: TimeInterval = 1.0,
                onChange: @escaping (String, StatusLineSnapshot) -> Void) {
        self.dir = dir
        self.interval = interval
        self.onChange = onChange
    }

    public func snapshot(for tabId: String) -> StatusLineSnapshot? {
        queue.sync { snapshots[tabId] }
    }

    /// What a tab should show until Claude renders in it — which is after its first output, so
    /// without this the row appears seconds late and changes the terminal's height while the user
    /// is already typing. The remembered snapshot for this directory, or at least its path.
    public func initialSnapshot(cwd: String) -> StatusLineSnapshot? {
        guard !cwd.isEmpty else { return nil }
        return queue.sync {
            var base = readRememberedSnapshot(cwd: cwd, dir: dir) ?? {
                var empty = StatusLineSnapshot()
                empty.cwd = collapseHome(cwd)
                return empty
            }()
            base = applyingLimits(readRememberedLimits(dir: dir), to: base)
            return base
        }
    }

    public func start() {
        guard timer == nil else { return }
        let timer = DispatchSource.makeTimerSource(queue: queue)
        timer.schedule(deadline: .now(), repeating: interval)
        timer.setEventHandler { [weak self] in self?.scan() }
        timer.resume()
        self.timer = timer
    }

    public func stop() {
        timer?.cancel()
        timer = nil
    }

    /// Forgets a tab's snapshot and removes its file — a closed tab's numbers must not reappear
    /// under a later tab that happens to reuse the id.
    public func forget(_ tabId: String) {
        queue.async {
            self.seen.removeValue(forKey: tabId)
            self.snapshots.removeValue(forKey: tabId)
            try? FileManager.default.removeItem(atPath: "\(self.dir)/\(tabId).json")
        }
    }

    // MARK: - The remembered layer

    /// Keeps what this snapshot knows and another one may not: the account's limits, and the last
    /// state of this directory. A snapshot without limits leaves the remembered ones alone.
    private func remember(_ snapshot: StatusLineSnapshot) {
        // Never backwards. The first scan reads every file in the directory, in whatever order
        // the file system hands them over — without this the oldest one could be the one that
        // ends up remembered, and a fresh tab would start from a window that closed hours ago.
        if let limits = limitFields(of: snapshot),
           limits.updatedAt >= (readRememberedLimits(dir: dir)?.updatedAt ?? 0) {
            write(limits, to: "limits.json")
        }
        // No window size means Claude has not really rendered yet — remembering that would hand
        // the next tab a scale of zero.
        if let cwd = snapshot.cwd, !cwd.isEmpty, snapshot.totalTokens > 0,
           snapshot.updatedAt >= (readRememberedSnapshot(cwd: cwd, dir: dir)?.updatedAt ?? 0) {
            write(snapshot, to: "\(cwdKey(cwd)).json")
        }
    }

    /// Through a temporary file and a rename, like `writeSnapshot` — a half-written file must
    /// never be what another window reads.
    private func write<T: Encodable>(_ value: T, to name: String) {
        guard let data = try? JSONEncoder().encode(value) else { return }
        try? FileManager.default.createDirectory(atPath: lastDir, withIntermediateDirectories: true,
                                                 attributes: [.posixPermissions: 0o700])
        let temp = "\(lastDir)/.\(name).tmp"
        guard (try? data.write(to: URL(fileURLWithPath: temp), options: .atomic)) != nil
        else { return }
        _ = try? FileManager.default.removeItem(atPath: "\(lastDir)/\(name)")
        try? FileManager.default.moveItem(atPath: temp, toPath: "\(lastDir)/\(name)")
    }

    private func scan() {
        guard let entries = try? FileManager.default.contentsOfDirectory(atPath: dir) else { return }
        for entry in entries where entry.hasSuffix(".json") {
            let tabId = String(entry.dropLast(5))
            let path = "\(dir)/\(entry)"
            guard let attributes = try? FileManager.default.attributesOfItem(atPath: path),
                  let modified = attributes[.modificationDate] as? Date else { continue }
            if let last = seen[tabId], last >= modified { continue }
            guard let written = readSnapshot(dir: dir, tabId: tabId) else { continue }

            // What the producer could not know is filled in here: Claude Code omits the rate
            // limits until a request has been made, so the first snapshot after `--continue`
            // would otherwise blank the Session and Week row it had a moment ago.
            let snapshot = resolvedSnapshot(written, dir: dir)
            seen[tabId] = modified
            snapshots[tabId] = snapshot
            remember(written)
            DispatchQueue.main.async { self.onChange(tabId, snapshot) }
        }
    }
}
